### PR TITLE
Bvh structure

### PR DESCRIPTION
**Pull request by alexd2580**

### CODE REVIEW

File: src/bvh.rs [hunk 1]
```diff
@@ -0,0 +1,508 @@
...
+//! [`BVH`]: struct.BVH.html
+//!
+
+use aabb::{AABB, Bounded};
+use ray::Ray;
+use std::boxed::Box;
+use std::f32;
+use std::iter::repeat;
+
+/// Enum which describes the union type of a node in a [`BVH`]s.
```
Comment (alexd2580) [inline src/bvh.rs]: Lexical error at line end.

Comment (alexd2580) [inline src/bvh.rs]: Corected in 16e4814.

File: src/bvh.rs [hunk 2]
```diff
@@ -0,0 +1,508 @@
...
+                aabb: a.aabb.union(&b.aabb),
+            }
+        }
+
+        // Create twelve buckets, and twelve index assignment vectors
+        const NUM_BUCKETS: usize = 6;
+        let mut buckets = [Bucket::empty(); NUM_BUCKETS];
+        let mut bucket_assignments: [Vec<usize>; NUM_BUCKETS] = Default::default();
+
+        // Iterate through all shapes
```
Comment (svenstaro) [inline src/bvh.rs]: ...and do what? Please describe the point of this loop.

File: src/bvh.rs [hunk 3]
```diff
@@ -0,0 +1,508 @@
...
+            child_r_aabb: child_r_aabb,
+            child_r: Box::new(BVHNode::build(shapes, child_r_indices)),
+        }
+    }
+
+    /// Prints a textual representation of the recursive [`BVH`] structure.
+    ///
+    /// [`BVH`]: struct.BVH.html
+    ///
+    fn print(&self, depth: usize) {
```
Comment (svenstaro) [inline src/bvh.rs]: Rename this to `pretty_print`.

File: src/bvh.rs [hunk 4]
```diff
@@ -0,0 +1,508 @@
...
+        }
+    }
+
+    /// Traverses the [`BVH`] recursively and returns a [`Vec`]
+    /// of all shapes which are hit with a high probability.
+    ///
+    /// [`BVH`]: struct.BVH.html
+    /// [`Vec`]: https://doc.rust-lang.org/std/vec/struct.Vec.html
+    ///
+    pub fn traverse_recursive(&self, ray: &Ray, indices: &mut Vec<usize>) {
```
Comment (svenstaro) [inline src/bvh.rs]: Document the arguments.

File: src/bvh.rs (hunk 3 above)
Comment (alexd2580) [inline src/bvh.rs]: Renamed in e5185b2.

File: src/bvh.rs (hunk 4 above)
Comment (alexd2580) [inline src/bvh.rs]: Partially fixed in e5185b2.

File: src/bvh.rs [hunk 5]
```diff
@@ -0,0 +1,508 @@
...
+    pub fn print(&self) {
+        self.root.print(0);
+    }
+
+    /// Traverses the tree recursively. Returns an array of all shapes, the [`AABB`]s of which
+    /// were hit.
+    ///
+    /// [`AABB`]: ../aabb/struct.AABB.html
+    ///
+    pub fn traverse_recursive<'a, T: Bounded>(&'a self, ray: &Ray, shapes: &'a [T]) -> Vec<&T> {
```
Comment (svenstaro) [inline src/bvh.rs]: Please document the parameters in some way.

Comment (alexd2580) [inline src/bvh.rs]: Fixed in b5c6d49.

File: src/bvh.rs (hunk 2 above)
Comment (alexd2580) [inline src/bvh.rs]: Fixed in b5c6d49.

File: src/lib.rs [hunk 6]
```diff
@@ -1,3 +1,6 @@
+//! A crate which exports rays, axis-aligned bounding boxes, and binary bounding
+//! volume hierarchies.
+
```
Comment (svenstaro) [inline src/lib.rs]: Please add a general usage example.

**Comment by svenstaro** [comment]
Great job! Merging.

**Pull request merged**